File: backend/src/services/voicevoxTextToSpeech.ts
import axios from 'axios';
import { inject, injectable } from 'tsyringe';
import { ITextToSpeech } from '../interfaces/ITextToSpeech';

@injectable()
class VoicevoxTextToSpeech implements ITextToSpeech {

    private baseUrl = process.env.VOICEVOX_BASE_URL;

    public async synthesize(text: string): Promise<Buffer> {

        const baseUrl = process.env.VOICEVOX_BASE_URL;

        // クエリ生成
        const query = await axios.post(`${this.baseUrl}/audio_query?speaker=1&text=${encodeURIComponent(text)}`).catch(error => {
            throw new Error('The request to the audio_query API failed. message: ' + error.message)
        });

        // 音声合成
        const audio = await axios.post(`${this.baseUrl}/synthesis?speaker=1`, JSON.stringify(query.data), {
            responseType: 'arraybuffer',
            headers: {
                'accept': 'audio/wav',
                'content-type': 'application/json'
            }
        }).catch(error => {
            throw new Error('The request to the synthesis API failed. message: ' + error.message)
        });

        return Buffer.from(audio.data, 'binary');
    }
}

export { VoicevoxTextToSpeech };

File: frontend/src/voicevox_text_to_speech.py
# import os
# import pyaudio
# import requests
# from abstract.abstract_text_to_speech import AbstractTextToSpeech

# class VoicevoxTextToSpeech(AbstractTextToSpeech):
#     def __init__(self) -> None:
#         self.base_url = os.getenv('VOICEVOX_BASE_URL')
#         self.speaker = os.getenv('VOICEVOX_SPEAKER')
    
#     def synthesize(self, text: str) -> None:
        
#         # クエリ作成
#         query = self.__post_audio_query(text)
        
#         # 音声合成
#         audio_data = self.__post_synthesis(query)
        
#         # 再生
#         self.__play_voice_vox_audio(audio_data)
        
#         return
    
#     def __post_audio_query(self, text) -> dict:
#         url = os.path.join(self.base_url, 'audio_query')
        
#         params = {
#             'speaker': self.speaker,
#             'text': text
#         }
        
#         try:
#             response = requests.post(url, params=params)
#             response.raise_for_status()
        
#         except Exception as err:
#             print(f'post request to voicevox query api failed: {err}')
#             raise

#         return response.json()
        
#     def __post_synthesis(self, query: dict) -> bytes:
        
#         url = os.path.join(self.base_url, 'synthesis')
        
#         params = {
#             'speaker': self.speaker
#         }
        
#         headers = {
#             'accept': 'audio/wav',
#             'content-type': 'application/json'
#         }
        
#         try:
#             response = requests.post(url, params=params, headers=headers, json=query)
#             response.raise_for_status()
            
#         except Exception as err:
#             print(f'post request to voicevox synthesis api failed: {err}')
#             raise
        
#         return response.content
        
#     def __play_voice_vox_audio(self, audio_data: bytes):

#         p = pyaudio.PyAudio()
        
#         stream = p.open(
#             format=pyaudio.paInt16,
#             channels=1,
#             rate=24000, # voicevoxは24000
#             output=True
#         )
        
#         stream.write(audio_data)
        
#         stream.stop_stream()
#         stream.close()
#         p.terminate

